File: src/main/java/fr/sunderia/bomberman/Bomberman.java
package fr.sunderia.bomberman;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.lang.reflect.Field;
import java.net.URI;
import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Random;
import java.util.UUID;
import java.util.stream.IntStream;

import org.apache.commons.codec.digest.DigestUtils;
import org.jglrxavpok.hephaistos.nbt.NBTException;
import org.jglrxavpok.hephaistos.nbt.NBTReader;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import fr.sunderia.bomberman.Structure.BlockPos;
import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
import net.kyori.adventure.key.Key;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.JoinConfiguration;
import net.kyori.adventure.text.format.NamedTextColor;
import net.kyori.adventure.title.TitlePart;
import net.minestom.server.MinecraftServer;
import net.minestom.server.attribute.Attribute;
import net.minestom.server.attribute.AttributeModifier;
import net.minestom.server.coordinate.Pos;
import net.minestom.server.coordinate.Vec;
import net.minestom.server.entity.Entity;
import net.minestom.server.entity.EntityType;
import net.minestom.server.entity.GameMode;
import net.minestom.server.entity.Player;
import net.minestom.server.event.Event;
import net.minestom.server.event.EventNode;
import net.minestom.server.event.item.PickupItemEvent;
import net.minestom.server.event.player.PlayerBlockPlaceEvent;
import net.minestom.server.event.player.PlayerChatEvent;
import net.minestom.server.event.player.PlayerDeathEvent;
import net.minestom.server.event.player.PlayerLoginEvent;
import net.minestom.server.event.player.PlayerSpawnEvent;
import net.minestom.server.extensions.Extension;
import net.minestom.server.extras.lan.OpenToLAN;
import net.minestom.server.instance.Instance;
import net.minestom.server.instance.InstanceContainer;
import net.minestom.server.instance.InstanceManager;
import net.minestom.server.instance.batch.AbsoluteBlockBatch;
import net.minestom.server.instance.batch.ChunkBatch;
import net.minestom.server.instance.block.Block;
import net.minestom.server.item.ItemStack;
import net.minestom.server.item.Material;
import net.minestom.server.network.packet.server.play.SetCooldownPacket;
import net.minestom.server.resourcepack.ResourcePack;
import net.minestom.server.timer.TaskSchedule;
import net.minestom.server.utils.NamespaceID;
import net.minestom.server.utils.chunk.ChunkUtils;
import net.minestom.server.world.DimensionType;

public class Bomberman extends Extension {

    static final Map<UUID, Integer> powerMap = new HashMap<>();
    private final Gson gson = new GsonBuilder().create();
    public static final Random random = new Random();
    private String resourcePackSha1;
    private final String resourcePackURL = "https://raw.githubusercontent.com/Sunderia/Bomberman/main/bomberman.zip";

    @Override
    public void initialize() {
        InstanceManager manager = MinecraftServer.getInstanceManager();
        InstanceContainer container = createInstanceContainer(manager);
        var extensionNode = getEventNode();
        OpenToLAN.open();
        registerListeners(extensionNode, container);
        String sha1 = null;
        try {
            ByteArrayOutputStream os = new ByteArrayOutputStream();
            URI.create(resourcePackURL).toURL().openStream().transferTo(os);
            sha1 = new String(DigestUtils.getSha1Digest().digest(os.toByteArray()));
        } catch(IOException e) {
            getLogger().error(Component.text("Couldn't get the SHA1 of the Resource Pack"), e);
        }
        this.resourcePackSha1 = sha1;
    }

    private void registerListeners(EventNode<Event> extensionNode, InstanceContainer container) {     
        extensionNode.addListener(PlayerLoginEvent.class, event -> event.setSpawningInstance(container));
        extensionNode.addListener(PlayerSpawnEvent.class, e -> {
            Player player = e.getPlayer();
            player.setGameMode(GameMode.ADVENTURE);
            player.teleport(player.getPosition().withY(45));
            player.setRespawnPoint(player.getPosition().withY(45));
            player.getInventory().addItemStack(ItemStack.of(Material.TNT).withMeta(builder -> builder.canPlaceOn(Block.STONE, Block.BRICKS).build()));
            powerMap.put(player.getUuid(), 2);
            player.scheduler()
                .scheduleTask(() -> player.sendActionBar(Component.join(JoinConfiguration.separator(Component.text(" ")),
                    Component.text("\uE000").style(b -> b.font(Key.key("bomberman", "font"))),
                    Component.text(": " + powerMap.get(player.getUuid())))
                        .style(b -> b.font(Key.key("default")))),
                TaskSchedule.immediate(), TaskSchedule.tick(10));
            if(resourcePackSha1 != null) player.setResourcePack(ResourcePack.forced("https://raw.githubusercontent.com/Sunderia/Bomberman/main/bomberman.zip", null));
        });

        extensionNode.addListener(PlayerChatEvent.class, e -> {
            Optional<GameMode> g = Arrays.stream(GameMode.values()).filter(gm -> e.getMessage().toUpperCase().contains(gm.name())).findFirst();
            g.ifPresent(gamemode -> {
                e.setCancelled(true);
                e.getPlayer().setGameMode(gamemode);
            });
        });

        extensionNode.addListener(PlayerDeathEvent.class, event -> {
            Player player = event.getPlayer();
            player.setGameMode(GameMode.SPECTATOR);
            List<Player> playerAlives = player.getInstance().getPlayers().stream().filter(p -> p.getGameMode() == GameMode.ADVENTURE).toList();
            if(playerAlives.size() != 1) return;
            Player winner = playerAlives.get(0);
            player.getInstance().getPlayers().stream()
                .filter(p -> p.getGameMode() != GameMode.ADVENTURE).forEach(p -> {
                    if(p.isDead()) p.respawn();
                    p.sendMessage(Component.text(winner.getUsername() + " Won"));
                    p.teleport(new Pos(0, 45, 0));
                    p.setGameMode(GameMode.ADVENTURE);
                });
            winner.sendTitlePart(TitlePart.TITLE, Component.text("You won", NamedTextColor.GREEN));
            winner.teleport(new Pos(0, 45, 0));
            generateStructure(winner.getInstance());
            resetGame(winner.getInstance());
        });

        extensionNode.addListener(PlayerBlockPlaceEvent.class, e -> {
            if(e.getBlock().id() != Block.TNT.id()) return;
            e.setCancelled(true);
            Player player = e.getPlayer();
            if(player.getGameMode() != GameMode.ADVENTURE && player.getGameMode() != GameMode.CREATIVE) return;
            Block blockBelow = player.getInstance().getBlock(e.getBlockPosition().sub(0, 1, 0));
            if(blockBelow.id() != Block.STONE.id() || player.getInstance().getBlock(e.getBlockPosition().add(0, 1, 0)).isSolid()) return;
            if(Cooldown.isInCooldown(e.getPlayer().getUuid(), "tnt")) return;
            e.setCancelled(false);
            e.getPlayer().sendPacket(new SetCooldownPacket(Material.TNT.id(), 0));
            e.consumeBlock(false);
            e.setBlock(Block.BARRIER);
            if(player.getGameMode() == GameMode.ADVENTURE) {
                final int timeInSeconds = 1;
                Cooldown c = new Cooldown(e.getPlayer().getUuid(), "tnt", timeInSeconds);
                c.start();
                SetCooldownPacket packet = new SetCooldownPacket(Material.TNT.id(), timeInSeconds * 20);
                e.getPlayer().sendPacket(packet);
            }
            PrimedTntEntity tnt = new PrimedTntEntity(e.getPlayer());
            tnt.setInstance(e.getInstance(), e.getBlockPosition().add(0.5d, 0, 0.5d));
        });

        extensionNode.addListener(PickupItemEvent.class, event -> {
            if(!(event.getLivingEntity() instanceof Player player) || event.getItemStack().material().id() != Material.NAUTILUS_SHELL.id()) return;
            if(player.getGameMode() != GameMode.ADVENTURE) {
                event.setCancelled(true);
                return;
            }
            int customModelData = event.getItemStack().meta().getCustomModelData();
            Powerup powerup = Powerup.values()[customModelData - 1];
            powerup.getEffect().accept(player);
        });
    }
    
    private InstanceContainer createInstanceContainer(InstanceManager manager) {
        DimensionType fullBright = DimensionType.builder(NamespaceID.from("sunderia:full_bright")).ambientLight(2.0f).build();
        MinecraftServer.getDimensionTypeManager().addDimension(fullBright);
        InstanceContainer container = manager.createInstanceContainer(fullBright);
        container.setGenerator(unit -> {
            unit.modifier().fillHeight(0, 40, Block.STONE);
        });
        generateStructure(container);
        return container;
    }

    private void resetGame(Instance instance) {
        powerMap.replaceAll((k,v) -> 2);
        instance.getPlayers().forEach(p -> {
            p.clearEffects();
            UUID[] uuids = p.getAttribute(Attribute.MOVEMENT_SPEED).getModifiers().stream().map(AttributeModifier::getId).toArray(UUID[]::new);
            for(UUID uuid : uuids) {
                p.getAttribute(Attribute.MOVEMENT_SPEED).removeModifier(uuid);
            }
        });
        instance.getEntities().stream()
            .filter(e -> e.getEntityType().id() == EntityType.TNT.id() || e.getEntityType().id() == EntityType.ITEM.id())
            .forEach(Entity::remove);
    }

    private void generateStructure(Instance container) {
        Structure blocks = parseNBT();
        if(blocks == null) return;
        Pos startPos = new Pos(0, 0, 0).sub(blocks.size().div(2)).withY(40);
        AbsoluteBlockBatch batch = new AbsoluteBlockBatch();
        for(BlockPos blockPos : blocks.blocks()) {
            if(blockPos.block().isAir()) continue;
            if(blockPos.block().id() == Block.BRICKS.id() && random.nextInt(3) == 2) continue;
            batch.setBlock(startPos.add(blockPos.vec()), blockPos.block());
            if(blockPos.block().id() == Block.BRICKS.id()) {
                batch.setBlock(startPos.add(blockPos.vec().add(0, 1, 0)), Block.BARRIER);
            }
        }
        ChunkUtils.optionalLoadAll(container, getAffectedChunks(batch), null)
            .thenRun(() -> batch.apply(container, () -> batch.clear()));
    }

    private Structure parseNBT() {
        InputStream stream = Bomberman.class.getResourceAsStream("/bomberman.nbt");
        try(NBTReader reader = new NBTReader(stream)) {
        List<BlockPos> structure = new LinkedList<>();
            JsonObject nbt = gson.fromJson(reader.read().toSNBT(), JsonObject.class);
            JsonArray palettes = nbt.getAsJsonArray("palette");
            Block[] palette = IntStream.range(0, palettes.size())
                .mapToObj(palettes::get).map(JsonElement::getAsJsonObject)
                .map(obj -> obj.getAsJsonPrimitive("Name").getAsString())
                .map(Block::fromNamespaceId).toArray(Block[]::new);
            JsonArray blockArray = nbt.getAsJsonArray("blocks");
            blockArray.forEach(el -> {
                JsonObject blockObj = el.getAsJsonObject();
                JsonArray jsonPos = blockObj.getAsJsonArray("pos");
                structure.add(new BlockPos(
                    new Vec(jsonPos.get(0).getAsInt(), jsonPos.get(1).getAsInt(), jsonPos.get(2).getAsInt()),
                    palette[blockObj.get("state").getAsInt()])
                );
            });
            int[] size = new int[3];
            JsonArray jsonSize = nbt.getAsJsonArray("size");
            for(int i = 0; i < 3; i++) size[i] = jsonSize.get(i).getAsInt();
            return new Structure(new Vec(jsonSize.get(0).getAsInt(), jsonSize.get(1).getAsInt(), jsonSize.get(2).getAsInt()), structure);
        } catch(IOException | NBTException e) {
            e.printStackTrace();
        }
        return null;
    }

    public static long[] getAffectedChunks(AbsoluteBlockBatch batch) {
        try {
            Field field = batch.getClass().getDeclaredField("chunkBatchesMap");
            field.setAccessible(true);

            Long2ObjectMap<ChunkBatch> chunkBatchesMap = (Long2ObjectMap<ChunkBatch>) field.get(batch);

            return chunkBatchesMap.keySet().toLongArray();

        } catch (NoSuchFieldException | IllegalAccessException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void terminate() {}    
}

File: src/main/kotlin/fr/sunderia/bomberman/Bomberman.kt
package fr.sunderia.bomberman

import com.google.gson.Gson
import fr.sunderia.bomberman.InstanceCreator.Companion.createInstanceContainer
import fr.sunderia.bomberman.party.*
import fr.sunderia.bomberman.utils.NPC
import fr.sunderia.bomberman.utils.ResourceUtils
import net.kyori.adventure.key.Key
import net.kyori.adventure.text.format.NamedTextColor
import net.minestom.server.MinecraftServer
import net.minestom.server.coordinate.Point
import net.minestom.server.coordinate.Pos
import net.minestom.server.coordinate.Vec
import net.minestom.server.entity.Entity
import net.minestom.server.entity.EntityType
import net.minestom.server.entity.metadata.display.ItemDisplayMeta
import net.minestom.server.extras.lan.OpenToLAN
import net.minestom.server.extras.velocity.VelocityProxy
import net.minestom.server.instance.Instance
import net.minestom.server.instance.InstanceContainer
import net.minestom.server.item.ItemComponent
import net.minestom.server.item.ItemStack
import net.minestom.server.item.Material
import net.minestom.server.item.component.HeadProfile
import net.minestom.server.registry.DynamicRegistry
import net.minestom.server.scoreboard.TeamBuilder
import net.minestom.server.utils.NamespaceID
import net.minestom.server.world.DimensionType
import org.apache.commons.codec.digest.DigestUtils
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.net.URI
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.util.*
import java.util.logging.Level
import java.util.logging.Logger
import kotlin.collections.Map

fun main() {
    val game = Bomberman()
    if(System.getenv().containsKey("VELOCITY_SECRET")) VelocityProxy.enable(System.getenv("VELOCITY_SECRET"))
    val mc = MinecraftServer.init()
    game.initialize()
    mc.start("0.0.0.0", 25565)
}

class Bomberman {

    companion object {
        val gson = Gson()
        val logger: Logger = Logger.getLogger("Bomberman")

        val fullBrightType = DimensionType.builder().ambientLight(2.0f).build()
        lateinit var fullbright: DynamicRegistry.Key<DimensionType>

        val fontKey = Key.key("bomberman", "font")
        lateinit var resourcePackSha1: String

        val powerMap = mutableMapOf<UUID, Int>()
        private lateinit var maps: Map<String, GameMap>
        private lateinit var lobbyInstance: Instance

        fun getLobbyInstance(): Instance {
            return lobbyInstance
        }
        fun getAvailableMaps(): Map<String, GameMap> = maps
        fun getMaxAmountOfPlayers(): Int = maps.values.maxOf { it.settings.maxPlayers }
    }

    private fun createPlayerPart(model: String, profile: HeadProfile, pos: Pos, translation: Point, lobby: Instance) {
        val entity = Entity(EntityType.ITEM_DISPLAY)
        entity.setNoGravity(true)
        val meta = entity.entityMeta as ItemDisplayMeta
        meta.displayContext = ItemDisplayMeta.DisplayContext.THIRD_PERSON_RIGHT_HAND
        meta.viewRange = .6f
        meta.leftRotation = floatArrayOf(.0f, .0f, .0f, 1f)
        meta.rightRotation = floatArrayOf(.0f, .0f, .0f, 1f)
        meta.scale = Vec(1.0, 1.0, 1.0)
        meta.translation = translation
        val headItem = ItemStack.of(Material.PLAYER_HEAD).withItemModel("player_display:player/${model}").with(ItemComponent.PROFILE, profile)
        meta.itemStack = headItem
        entity.setInstance(lobby, pos)
    }

    private fun createFakeNPC(lobby: Instance) {
        val map = mapOf(
            "head" to (Pos(.0, 1.4, .0) to Pos.ZERO), "torso" to (Pos(.0, 1.4, .0) to Pos(.0, -3072.0, .0)),
            "right_arm" to (Pos(.0, 1.4, .0) to Pos(.0, -1024.0, .0)), "left_arm" to (Pos(.0, 1.4, .0) to Pos(.0, -2048.0, .0)),
            "right_leg" to (Pos(.0, .7, .0) to Pos(.0, -4096.0, .0)), "left_leg" to (Pos(.0, .7, .0) to Pos(.0, -5120.0, .0)))

        for ((key, value) in map) {
            val (relativePos, transform) = value
            createPlayerPart(key, HeadProfile(NPC.BOMBERMAN_SKIN), Pos(5.0, 40.0, 10.0).add(relativePos), transform, lobby)
        }
    }

    fun initialize() {
        val uri = Bomberman::class.java.getResource("/maps/")?.toURI() ?: throw IOException("Could not find maps folder")
        val paths = ResourceUtils.getPaths(uri)
        logger.info("Found ${paths.size} maps")
        maps = paths.map {
            val name = it.fileName.toString().replace(".json", "")
            logger.info("Loading map $name")
            name to GameMap(name)
        }.toList().toMap()
        val manager = MinecraftServer.getInstanceManager()
        fullbright = MinecraftServer.getDimensionTypeRegistry().register(NamespaceID.from("sunderia:full_bright"), fullBrightType)
        val lobbyContainer: InstanceContainer = createInstanceContainer(manager)
        lobbyInstance = lobbyContainer
        NPC.createNPC(lobbyContainer)
        createFakeNPC(lobbyContainer)
        OpenToLAN.open()
        //MojangAuth.init()
        Listeners.registerListeners(lobbyContainer)
        val commandManager = MinecraftServer.getCommandManager()
        commandManager.register(PartyCommand())
        commandManager.register(GameCommand())

        PrimedTntEntity.pierceTeam = TeamBuilder("Pierce TNT", MinecraftServer.getTeamManager()).teamColor(NamedTextColor.AQUA).build()

        try {
            ByteArrayOutputStream().use { os ->
                URI.create("https://raw.githubusercontent.com/Sunderia/Bomberman/main/bomberman.zip").toURL().openStream().use { input ->
                    input.transferTo(os)
                    resourcePackSha1 = String(DigestUtils.getSha1Digest().digest(os.toByteArray()))
                }
            }
        } catch (e: IOException) {
            logger.log(Level.SEVERE, e) { "Couldn't get the SHA1 of the Resource Pack" }
        }
        logger.info("Bomberman starting")
    }
}